### PR TITLE
Move Location of Experiments Call to CD Internal

### PR DESCRIPTION
**Pull request by amitla1**
The call needs to be after snapshots are created.

### CODE REVIEW

**Review by melotic — CHANGES_REQUESTED** [review]
This breaks experiments in CD. In CD these are written to disk, and then they are submitted to CG in 1P uses. 

You need a way to process experiments at different times.

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/IExperimentService.cs [hunk 1]
```diff
@@ -22,5 +22,5 @@ public interface IExperimentService
     /// Called when all detectors have finished executing. Processes the experiments and reports the results.
     /// </summary>
     /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Experiments/IExperimentService.cs:24]: nit: add documentation on the new flag

File: src/Microsoft.ComponentDetection.Orchestrator/Orchestrator.cs [hunk 2]
```diff
@@ -29,6 +29,8 @@ namespace Microsoft.ComponentDetection.Orchestrator;
 
 public class Orchestrator
 {
+    public const bool AutomaticallyProcessExperiments = false;
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Orchestrator.cs:32]: This needs to belong in [`DetectorExperiments`](https://github.com/microsoft/component-detection/blob/4698b588479d3f7fdf9759666f4cbb30e28eb4ad/src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs)

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/ExperimentService.cs [hunk 3]
```diff
@@ -109,13 +109,18 @@ private void FilterExperiments(IComponentDetector detector, int count)
...
     /// <inheritdoc />
-    public async Task FinishAsync()
+    public async Task FinishAsync(bool shouldCheckAutomaticProcessFlag = false)
     {
         if (!DetectorExperiments.AreExperimentsEnabled)
         {
             return;
         }
 
+        if (!shouldCheckAutomaticProcessFlag && !DetectorExperiments.AutomaticallyProcessExperiments)
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Experiments/ExperimentService.cs:119]: Can we add a test for this?

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs [hunk 4]
```diff
@@ -7,6 +7,8 @@ namespace Microsoft.ComponentDetection.Orchestrator.Experiments;
 /// </summary>
 public static class DetectorExperiments
 {
+    public const bool AutomaticallyProcessExperiments = false;
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs:10]: shouldnt this be true?

**Comment by melotic** [comment]
Can you fix the build errors?

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs (hunk 4 above)
Comment (amitla1) [reply src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs:10]: no

File: test/Microsoft.ComponentDetection.Orchestrator.Tests/Services/DetectorProcessingServiceTests.cs [hunk 5]
```diff
@@ -520,7 +520,7 @@ public async Task ProcessDetectorsAsync_FinishesExperimentsAsync()
 
         await this.serviceUnderTest.ProcessDetectorsAsync(DefaultArgs, this.detectorsToUse, new DetectorRestrictions());
 
-        this.experimentServiceMock.Verify(x => x.FinishAsync(), Times.Once());
+        this.experimentServiceMock.Verify(x => x.FinishAsync(false), Times.Once());
```
Comment (melotic) [inline test/Microsoft.ComponentDetection.Orchestrator.Tests/Services/DetectorProcessingServiceTests.cs:523]: This should be `It.Is<bool>(x => !x)`, not `false`

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs (hunk 4 above)
Comment (melotic) [reply src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs:10]: Your logic is:

```csharp
if (!shouldCheckAutomaticProcessFlag && !DetectorExperiments.AutomaticallyProcessExperiments)
{
    return;
}
```
In the `DetectorProcessingService`, `shouldCheckAutomaticProcessFlag` is set to `false`.
`AutomaticallyProcessExperiments` is `false`.

`!shouldCheckAutomaticProcessFlag && !DetectorExperiments.AutomaticallyProcessExperiments`
`!false && !false`
`true && true`
`true` => `FinishAsync` is a No-OP in `DetectorProcessingService`.

It should be true.

File: src/Microsoft.ComponentDetection.Orchestrator/Orchestrator.cs (hunk 2 above)
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Orchestrator.cs:32]: this should be deleted

File: test/Microsoft.ComponentDetection.Orchestrator.Tests/Services/DetectorProcessingServiceTests.cs [hunk 6]
```diff
@@ -520,6 +520,8 @@ public async Task ProcessDetectorsAsync_FinishesExperimentsAsync()
 
         await this.serviceUnderTest.ProcessDetectorsAsync(DefaultArgs, this.detectorsToUse, new DetectorRestrictions());
 
+        DetectorExperiments.AutomaticallyProcessExperiments = true;
```
Comment (melotic) [inline test/Microsoft.ComponentDetection.Orchestrator.Tests/Services/DetectorProcessingServiceTests.cs:523]: Remove this. It should be set automatically.

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs [hunk 7]
```diff
@@ -7,6 +7,11 @@ namespace Microsoft.ComponentDetection.Orchestrator.Experiments;
 /// </summary>
 public static class DetectorExperiments
 {
+    /// <summary>
+    /// Check to automatically proccess experiments.
+    /// </summary>
+    public static bool AutomaticallyProcessExperiments { get; set; }
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs:13]: This should be true

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/ExperimentService.cs [hunk 8]
```diff
@@ -116,6 +116,11 @@ public async Task FinishAsync()
             return;
         }
 
+        if (!DetectorExperiments.AutomaticallyProcessExperiments)
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Experiments/ExperimentService.cs:119]: Why did we remove the force flag?

File: src/Microsoft.ComponentDetection.Orchestrator/Services/DetectorProcessingService.cs [hunk 9]
```diff
@@ -122,6 +122,7 @@ public async Task<DetectorProcessingResult> ProcessDetectorsAsync(IDetectionArgu
             }).ToList();
 
         var results = await Task.WhenAll(scanTasks);
+        DetectorExperiments.AutomaticallyProcessExperiments = false;
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Services/DetectorProcessingService.cs:125]: Why??

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs [hunk 10]
```diff
@@ -7,6 +7,11 @@ namespace Microsoft.ComponentDetection.Orchestrator.Experiments;
 /// </summary>
 public static class DetectorExperiments
 {
+    /// <summary>
+    /// Check to automatically proccess experiments.
+    /// </summary>
+    public static bool AutomaticallyProcessExperiments => true;
```
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Experiments/DetectorExperiments.cs:13]: `=>` makes this a get only property.

File: src/Microsoft.ComponentDetection.Orchestrator/Experiments/ExperimentService.cs (hunk 8 above)
Comment (melotic) [inline src/Microsoft.ComponentDetection.Orchestrator/Experiments/ExperimentService.cs:119]: This should be unit tested.

**Review by melotic — APPROVED** [review]

**Pull request merged**